### ISSUE TITLE
Query :: invalid results returned for some groupjoin related scenarios

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFramework#5103]
Repro:

```cs
    public class Comment
    {
        public int CommentId { get; set; }
        public string Name { get; set; }
        public Post Post { get; set; }
    }

    public class Post
    {
        public int PostId { get; set; }
        public Blog Blog { get; set; }

        public List<Comment> Comments { get; set; }
    }

    public class Blog
    {
        public int BlogId { get; set; }

        public List<Post> Posts { get; set; }
    }

    public class MyContext : DbContext
    {
        public DbSet<Blog> Blogs { get; set; }
        public DbSet<Post> Posts { get; set; }
        public DbSet<Comment> Comments { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer("Server=.;Database=Repro5103;Trusted_Connection=True;MultipleActiveResultSets=true;");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();

                var c111 = new Comment { Name = "c111" };
                var c112 = new Comment { Name = "c112" };
                var c121 = new Comment { Name = "c121" };
                var c122 = new Comment { Name = "c122" };
                var c211 = new Comment { Name = "c211" };
                var c212 = new Comment { Name = "c212" };
                var c221 = new Comment { Name = "c221" };
                var c222 = new Comment { Name = "c222" };

                var post11 = new Post { Comments = new List<Comment> { c111, c112 } };
                var post12 = new Post { Comments = new List<Comment> { c121, c122 } };
                var post21 = new Post { Comments = new List<Comment> { c211, c212 } };
                var post22 = new Post { Comments = new List<Comment> { c221, c222 } };

                var blog1 = new Blog { Posts = new List<Post> { post11, post12 } };
                var blog2 = new Blog { Posts = new List<Post> { post21, post22 } };

                ctx.Blogs.AddRange(blog1, blog2);
                ctx.Posts.AddRange(post11, post12, post21, post22);
                ctx.Comments.AddRange(c111, c112, c121, c122, c211, c212, c221, c222);
                ctx.SaveChanges();
            }

            using (var ctx = new MyContext())
            {
                var result = ctx.Comments
                    .Include(c => c.Post)
                    .Where(c => c.Post.Blog.BlogId == 1)
                    .ToList();
            }
        }
    }
```

results:
"111", "111", "121", "121"

expected:
"111", "112", "121", "122"

**Comment by maumar** [comment]
possibly related to https://github.com/aspnet/EntityFramework/pull/5048

**Comment by rowanmiller** [comment]
@mikary can you confirm if this is fixed

**Comment by mikary** [comment]
Confirmed this now produces the correct results. Appears to be related to materialization issue in #5179, which was fixed in e169d17243760ac3eece58bf8187905665ab67f6 and additional test coverage added in d913e7c2318fadf9bdd1e0ea4ce0426ede42d0e3
